### PR TITLE
Wasm pool tests

### PR DESCRIPTION
**Pull request by patrick-east**
This adds in some coverage of the wasm pool and vm instantiation code. In particular the focus is on the initialization of the data with each VM, ensuring it is propagated correctly to each VM in the pool.

First thing added is the missing implementation for RemoveDataPath on the pool... Somehow that slipped through the initial implementation PR, and until now there was no coverage in the automated tests.

One detail, which we may need to address sooner rather than later, is the cyclic import issue that these tests run into. They have need to build wasm modules, but the code paths in OPA that do that (compiler/rego packages) rely on code that imports the SDK... so the tests in the same `opa` package are unable to import them without getting a build issue due to the cycle. To prevent this temporarily I flagged the tests are only running when we have the "nop" implementation of the resolver active, but this means we need to run an additional test command.. which is pretty undesirable. To correct the issue we would need to do some substantial refactoring (or something else? I'm open to suggestions)

### CODE REVIEW

**Review by tsandall — COMMENTED** [review]
Here's the cyclic dependency:

```
package github.com/open-policy-agent/opa/internal/wasm/sdk/opa (test)
        imports github.com/open-policy-agent/opa/compile
        imports github.com/open-policy-agent/opa/rego
        imports github.com/open-policy-agent/opa/internal/bundle
        imports github.com/open-policy-agent/opa/resolver/wasm
        imports github.com/open-policy-agent/opa/internal/wasm/sdk/opa: import cycle not allowed in test
```

One option would be to just pass the resolvers or resolver factories into the `rego` package. This solution would require small changes to the subcommands, server, REPL, etc. but I think those would be relatively simple. Basically anywhere we use the `rego` package inside of OPA would need to get updated.

File: internal/wasm/sdk/opa/pool.go [hunk 1]
```diff
@@ -222,6 +222,40 @@ func (p *pool) SetDataPath(path []string, value interface{}) error {
 }
 
 func (p *pool) RemoveDataPath(path []string) error {
```
Comment (tsandall) [inline internal/wasm/sdk/opa/pool.go:224]: Are we sure this function as well as `SetDataPath` are safe for concurrent access? It looks like we're relying on the fine-grained locking from the other functions however the rest of the exported functions on the pool all use coarse-grained locking. I'm a bit concerned about concurrent access here.

**Comment by patrick-east** [comment]
> One option would be to just pass the resolvers or resolver factories into the rego package.

The main downside with that approach is that usage of the `Load` and `LoadBundle` options on the Rego objects would not automatically get wasm bundles to work. The caller would need to call loader API's to get the bundle first, extract/initialize the wasm stuff, then pass it off to the Rego API. It isn't the end of the world, but makes the Rego API's a little inconsistent.

**Comment by tsandall** [comment]
> The caller would need to call loader API's to get the bundle first, extract/initialize the wasm stuff, then pass it off to the Rego API. It isn't the end of the world, but makes the Rego API's a little inconsistent.

Can't the caller essentially pass a function that takes a loaded bundle and returns the resolver(s)? I would have thought we could implement that once for the wasm implementation and then the API would just be `rego.New(rego.Load(<path>), rego.Resolvers(r))`. The caller wouldn't need to do anything with the bundle content.

File: internal/wasm/sdk/opa/pool.go (hunk 1 above)
Comment (patrick-east) [reply internal/wasm/sdk/opa/pool.go:224]: As sure as I am about the rest of the sdk code 😅   Did you have some specific concern or issue you saw?

It's the same concurrency model as when the VM's get new data/policy.  The public `SetPolicyData` one uses the course locking to cover the initialization phase of the pool, and for getting a copy of the current set of data+policy (which the `activate` helper guards with the same lock while setting). The flow where we acquire each VM and update, or throw them away, remains the same. So if anything its consistently wrong 🤣 

I'm pretty confident that once we have acquire the VM from `wait` we are safe. The main points to be worried about are the interactions between `wait`, `remove`, and the normal `Acquire`/`Release` workflows. I can take some time to go through and audit those... we don't really have any test coverage for them.

**Comment by patrick-east** [comment]
> Can't the caller essentially pass a function that takes a loaded bundle and returns the resolver(s)? I would have thought we could implement that once for the wasm implementation and then the API would just be rego.New(rego.Load(<path>), rego.Resolvers(r)). The caller wouldn't need to do anything with the bundle content.

Yep, I still have concerns about the need for every client to pass in the same option to get bundles to work as expected (same inconsistency issue IMO) but it would avoid the import cycle and we can always iterate later so I'm game. Better that the build flag 👍

**Review by tsandall — APPROVED** [review]
LGTM

**Pull request merged**